### PR TITLE
#76

### PR DESCRIPTION
**Pull request by thaysvieira**
As a car providers can create coupons

### CODE REVIEW

File: carRental/src/main/java/com/amigoscode/chohort2/carRental/carProviderUser/CarProviderUserRepository.java [hunk 1]
```diff
@@ -1,9 +1,12 @@
 package com.amigoscode.chohort2.carRental.carProviderUser;
 
 import com.amigoscode.chohort2.carRental.abstracts.repository.CRJpaRepository;
-import com.amigoscode.chohort2.carRental.car.Car;
+import org.springframework.data.repository.query.Param;
 import org.springframework.stereotype.Repository;
 
+import java.util.Optional;
+
 @Repository
```
Comment (Esmaeeilenani) [inline carRental/src/main/java/com/amigoscode/chohort2/carRental/carProviderUser/CarProviderUserRepository.java:9]: extends JpaRepository

File: carRental/src/main/java/com/amigoscode/chohort2/carRental/carProviderCoupon/CarProviderCouponService.java [hunk 2]
```diff
@@ -1,11 +1,34 @@
...
+                .setCouponCode(carProviderCouponVM.getCouponCode())
+                .setStartDate(carProviderCouponVM.getStartDate())
+                .setEndDate(carProviderCouponVM.getEndDate())
+                .setNumOfUsePerUser(carProviderCouponVM.getNumOfUsePerUser())
+                .setIsAvailable(carProviderCouponVM.getIsAvailable());
+        carProviderCouponRepository.save(carProviderCoupon);
+    }
+    private Long getCarProvideId() {
+        User user = userService.getLoggedInUser();
+        Optional<CarProviderUser> carProviderUserOptional = carProviderUserRepository.findByUserId(user.getId());
```
Comment (Esmaeeilenani) [inline carRental/src/main/java/com/amigoscode/chohort2/carRental/carProviderCoupon/CarProviderCouponService.java:31]: you should make sure its not null

File: carRental/src/main/java/com/amigoscode/chohort2/carRental/carProviderCoupon/VM/CarProviderCouponVM.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import lombok.Setter;
+import lombok.experimental.Accessors;
+
+import java.time.LocalDate;
+
+@Setter
+@Getter
+@Accessors(chain = true)
+public class CarProviderCouponVM {
+    @NotEmpty
```
Comment (DmitriKonnovNN) [inline carRental/src/main/java/com/amigoscode/chohort2/carRental/carProviderCoupon/VM/CarProviderCouponVM.java:15]: VM classes should implement Serializable;
Accordint to Java Object Serialization Specification it is highly recommended "that all serializable classes other than enum types explicitly declare serialVersionUID values".

**Review by DmitriKonnovNN — CHANGES_REQUESTED** [review]

File: carRental/src/main/java/com/amigoscode/chohort2/carRental/carProviderCoupon/CarProviderCouponController.java [hunk 4]
```diff
@@ -1,12 +1,28 @@
...
 @RequiredArgsConstructor
 public class CarProviderCouponController {
+    private final Logger log = LoggerFactory.getLogger(CarProviderCouponController.class);
+
     private final CarProviderCouponService carProviderCouponService;
+    @PostMapping("/coupons")
+    @Secured({AuthorityConstants.CAR_PROVIDER})
+    public ResponseEntity<Void> saveCarProviderCoupon( @RequestBody CarProviderCouponVM carProviderCouponVM){
+        log.info("car provider coupon info {}", carProviderCouponVM);
+        carProviderCouponService.save(carProviderCouponVM);
```
Comment (Esmaeeilenani) [inline carRental/src/main/java/com/amigoscode/chohort2/carRental/carProviderCoupon/CarProviderCouponController.java:24]: return the create coupon object

File: carRental/src/main/java/com/amigoscode/chohort2/carRental/carProviderCoupon/CarProviderCouponService.java [hunk 5]
```diff
@@ -1,11 +1,33 @@
...
+                .setCouponCode(carProviderCouponVM.getCouponCode())
+                .setStartDate(carProviderCouponVM.getStartDate())
+                .setEndDate(carProviderCouponVM.getEndDate())
+                .setNumOfUsePerUser(carProviderCouponVM.getNumOfUsePerUser())
+                .setIsAvailable(carProviderCouponVM.getIsAvailable());
+        carProviderCouponRepository.save(carProviderCoupon);
+    }
+    private Long getCarProvideId() {
+        User user = userService.getLoggedInUser();
+        var carProviderUserOptional = carProviderUserRepository.findByUserId(user.getId()).orElseThrow(()-> new NullPointerException(ErrorConstants.INTERNAL));
```
Comment (Esmaeeilenani) [inline carRental/src/main/java/com/amigoscode/chohort2/carRental/carProviderCoupon/CarProviderCouponService.java:30]: use APIException class

File: carRental/src/main/java/com/amigoscode/chohort2/carRental/carProviderUser/CarProviderUserRepository.java [hunk 6]
```diff
@@ -1,9 +1,11 @@
...
-import com.amigoscode.chohort2.carRental.car.Car;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.repository.query.Param;
 import org.springframework.stereotype.Repository;
 
+import java.util.Optional;
+
 @Repository
-public interface CarProviderUserRepository extends CRJpaRepository<CarProviderUser,Long> {
+public interface CarProviderUserRepository extends JpaRepository<CarProviderUser,Long> {
```
Comment (Esmaeeilenani) [inline carRental/src/main/java/com/amigoscode/chohort2/carRental/carProviderUser/CarProviderUserRepository.java:9]: no need for @Param

File: carRental/src/main/java/com/amigoscode/chohort2/carRental/carProviderCoupon/CarProviderCouponRepository.java [hunk 7]
```diff
@@ -1,6 +1,8 @@
 package com.amigoscode.chohort2.carRental.carProviderCoupon;
 
 import com.amigoscode.chohort2.carRental.abstracts.repository.CRJpaRepository;
+import org.springframework.stereotype.Repository;
 
+@Repository
 public interface CarProviderCouponRepository extends CRJpaRepository <CarProviderCoupon, Long> {
```
Comment (Esmaeeilenani) [inline carRental/src/main/java/com/amigoscode/chohort2/carRental/carProviderCoupon/CarProviderCouponRepository.java:7]: extends JpaRepository

File: carRental/src/main/java/com/amigoscode/chohort2/carRental/carProviderCoupon/CarProviderCouponController.java [hunk 8]
```diff
@@ -1,12 +1,28 @@
...
 public class CarProviderCouponController {
+    private final Logger log = LoggerFactory.getLogger(CarProviderCouponController.class);
+
     private final CarProviderCouponService carProviderCouponService;
+    @PostMapping("/coupons")
+    @Secured({AuthorityConstants.CAR_PROVIDER})
+    public ResponseEntity<CarProviderCouponVM> saveCarProviderCoupon( @RequestBody CarProviderCouponVM carProviderCouponVM){
+        log.info("car provider coupon info {}", carProviderCouponVM);
+        carProviderCouponService.save(carProviderCouponVM);
+        return ResponseEntity.accepted().body(carProviderCouponVM);
```
Comment (Esmaeeilenani) [inline carRental/src/main/java/com/amigoscode/chohort2/carRental/carProviderCoupon/CarProviderCouponController.java:25]: you should return the created on not what is coming from the client

**Review by Esmaeeilenani — APPROVED** [review]

**Review by DmitriKonnovNN — APPROVED** [review]
LGTM

**Pull request merged**